db/compaction/compaction_job_internal.h

```diff
@@ -0,0 +1,395 @@
+// Data structures used for compaction_job and compaction_service_job
+namespace ROCKSDB_NAMESPACE {
+
+// Files produced by this subcompaction
+struct Output {
+  Output(FileMetaData&& _meta, const InternalKeyComparator& _icmp,
+         bool _enable_order_check, bool _enable_hash, bool _finished = false,
+         uint64_t precalculated_hash = 0)
+      : meta(std::move(_meta)),
+        validator(_icmp, _enable_order_check, _enable_hash, precalculated_hash),
+        finished(_finished) {}
+  FileMetaData meta;
+  OutputValidator validator;
+  bool finished;
+  std::shared_ptr<const TableProperties> table_properties;
+};
+
+struct CompactionOutputs {
+ public:
+  Output* current_output() {
+    // This subcompaction's output could be empty if compaction was aborted
+    // before this subcompaction had a chance to generate any output files.
+    // When subcompactions are executed sequentially this is more likely and
+    // will be particularly likely for the later subcompactions to be empty.
+    // Once they are run in parallel however it should be much rarer.
+    if (outputs.empty()) {
+      return nullptr;
+    } else {
+      return &outputs.back();
+    }
+  }
+
+  Status AddToBuilder(const Slice& key, const Slice& value);
+
+  Status ProcessOutFlowIfNeeded(const Slice& key, const Slice& value) const {
+    if (!blob_garbage_meter) {
+      return Status::OK();
+    }
+    return blob_garbage_meter->ProcessOutFlow(key, value);
+  }
+
+  Slice SmallestUserKey() const {
+    if (!outputs.empty() && outputs[0].finished) {
+      return outputs[0].meta.smallest.user_key();
+    } else {
+      return Slice{nullptr, 0};
+    }
+  }
+
+  Slice LargestUserKey() const {
+    if (!outputs.empty() && outputs.back().finished) {
+      return outputs.back().meta.largest.user_key();
+    } else {
+      return Slice{nullptr, 0};
+    }
+  }
+
+  void RemoveLastEmptyOutput() {
+    if (!outputs.empty() && !outputs.back().meta.fd.file_size) {
+      // An error occurred, so ignore the last output.
+      outputs.pop_back();
+    }
+  }
+
+  void Cleanup() {
+    if (builder != nullptr) {
+      // May happen if we get a shutdown call in the middle of compaction
+      builder->Abandon();
+      builder.reset();
+    }
+  }
+
+  Status AddRangeTombstones(Slice* comp_start, Slice* comp_end,
+                            CompactionRangeDelAggregator* range_del_agg,
+                            CompactionIterationStats* range_del_out_stats,
+                            bool bottommost_level,
+                            const InternalKeyComparator& icmp,
+                            SequenceNumber earliest_snapshot,
+                            const Slice* next_table_min_key);
+
+  std::vector<Output> outputs;
+  std::vector<BlobFileAddition> blob_file_additions;
+  std::unique_ptr<BlobGarbageMeter> blob_garbage_meter;
+  std::unique_ptr<WritableFileWriter> outfile;
+  std::unique_ptr<TableBuilder> builder;
+  uint64_t current_output_file_size = 0;
+};
```

Comment:
Should we try to separate this class to another file too?

Reply:
done.